plugins/usbpro/OpenDeckWidget.cpp

```diff
@@ -0,0 +1,79 @@
+/*
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * OpenDeckWidget.cpp
+ * The OpenDeck Widget.
+ * Copyright (C) 2021 Simon Newton, Igor Petrovic
+ */
+
+#include <vector>
+
+#include "ola/Constants.h"
+#include "plugins/usbpro/OpenDeckWidget.h"
+
+namespace ola {
+namespace plugin {
+namespace usbpro {
+
+/**
+ * OpenDeckWidget Constructor
+ */
+OpenDeckWidget::OpenDeckWidget(
+  ola::io::ConnectedDescriptor *descriptor)
+    : GenericUsbProWidget(descriptor) {
+}
+
+
+bool OpenDeckWidget::SendDMX(const DmxBuffer &data) {
+  if (data != internal_buffer) {
+    std::vector<uint8_t> send_buffer = {};
+
+    for (size_t index = 0; index < data.Size(); index++) {
+      if (data.Get(index) != internal_buffer.Get(index)) {
+        send_buffer.push_back((index+1) & 0xFF);
+        send_buffer.push_back((index+1) >> 8);
+        send_buffer.push_back(data.Get(index));
+      }
+    }
+
+    internal_buffer = data;
+
+    if (send_buffer.size() >= MAX_DIFF_CHANNELS) {
+      // Just send the full frame in this case
+      struct {
+        uint8_t start_code;
+        uint8_t dmx[DMX_UNIVERSE_SIZE];
+      } full_frame;
+
+      full_frame.start_code = DMX512_START_CODE;
+      unsigned int length = DMX_UNIVERSE_SIZE;
+      data.Get(full_frame.dmx, &length);
+
+      return SendMessage(DMX_LABEL,
+                         reinterpret_cast<uint8_t*>(&full_frame),
+                         length + 1);
+    } else {
+      return SendMessage(DMX_SLOT_VALUE_DIFF_LABEL,
+                         &send_buffer[0],
+                         send_buffer.size());
+    }
+  } else {
+    // Nothing new to send
```

Comment:
Please add an OLA_DEBUG here so we can see what's going on.

Will your system keep sending the old packet indefinitely?

What happens if the board is unplugged and replugged without OLA being restarted? Does it reconnect, does the internal buffer get cleared on disconnect?

Reply:
I'm not sure where should I handle the "disconnect" event in OLA for my device.

Reply:
But OLA_DEBUG is added

Reply:
Please can you test this workflow, I think the following should work:

1. Ideally have something/a few somethings connected to DMX output (e.g. LEDs)
2. Turn on channel 1 and confirm output
3. Wait until it's logging the nothing changed messages
4. Unplug OpenDeck
5. Turn on channel 2
6. Wait until it's logging the nothing changed messages
7. Replug device
8. Confirm channel 2 lights up (i.e. it doesn't continue to use the internal buffer where nothing has changed and hence doesn't send new data)
9. Turn on channel 3 (this should fix the issue in either case if it's reconnected okay)

If the above doesn't work as expected, then some reading of the -l 4 logs or testing my side should hopefully figure out where we need to fix it...

Reply:
I've just tried this. After the reconnection to USB it takes few seconds for OLA to find the device again, after which it immediately resends all values, and everything is back as it was before the disconnection. Was this the thing you were interested in?

Reply:
Also, changing the values of some channels once the device is unplugged results in those channels being correctly applied once the device is connected and detected by OLA again.